### PR TITLE
Added dart implementation

### LINKED ISSUE #123: Add Dart
Hi! I'd like to request support for the Dart language.

### PR DESCRIPTION
**Pull request by ricardoboss**
Fixes #123

### CODE REVIEW

File: dart/std_uri_template.dart [hunk 1]
```diff
@@ -0,0 +1,515 @@
...
+  }
+
+  static bool _isList(dynamic value) {
+    return value is List;
+  }
+
+  static bool _isMap(dynamic value) {
+    if (value is Map) {
+      // https://stackoverflow.com/a/173479/7898052
+      if (!const bool.fromEnvironment('dart.library.js')) {
```
Comment (andreaTP) [inline dart/std_uri_template.dart:352]: Is this the best we can do in Dart?

**Comment by ricardoboss** [comment]
In its current form, the dart file is not publishable as its own Dart package. That would require a `pubspec.yaml` with some metadata. Also, since [`pub.dev`](https://pub.dev) doesn't support namespaced packages, I didn't publish it as to leave the `std_uri_template` package name free for you to use.

**Comment by ricardoboss** [comment]
Maybe the `StdUriTemplate` class should be made `abstract` or have a private constructor to prevent instantiating it.

**Comment by ricardoboss** [comment]
> @andreaTP Do you mind sharing a good link on how to publish dart packages so that I can go through the required steps?

Dart is pretty well documented and you can find all you need here: https://dart.dev/tools/pub/publishing

File: dart/std_uri_template.dart (hunk 1 above)
Comment (ricardoboss) [reply dart/std_uri_template.dart:352]: The link is for PHP, that should be removed; my bad. I'll look into it, but it should work in dart.js environments.

**Comment by ricardoboss** [comment]
> @andreaTP I notice that you are re-allocating the StringBuffers while in (most of) the fully typed implementation we are "resetting" those for performance reasons.

The current implementation uses `token != null` to determine the current state. I could replace that with `token.isEmpty` and the allocations with `token.reset()`.

File: dart/std_uri_template.dart (hunk 1 above)
Comment (ricardoboss) [reply dart/std_uri_template.dart]: Dart has typed `Map`s as opposed to PHP. The simpler check should suffice since we don't have to explicitly check if the `Map` could be a `List` (i.e. only integer keys).

**Comment by ricardoboss** [comment]
I looked at the C# implementation and noticed they use more enums. I decided to stick with strings here, but enums may be better for performance. What is your preference?

**Comment by andreaTP** [comment]
> I looked at the C# implementation and noticed they use more enums. I decided to stick with strings here, but enums may be better for performance. What is your preference?

My personal preference is for enums, but it's not a requirement.
You probably started the implementation from a dynamically typed language while the Java implementation is commonly used as the reference.

File: dart/std_uri_template.dart (hunk 1 above)
Comment (andreaTP) [reply dart/std_uri_template.dart]: 👍

**Comment by andreaTP** [comment]
@ricardoboss I have reviewed your implementation and it looks like a solid starting point, thanks a lot!
Here is the list of what is left, in priority order:

- fix the failure in CI (I can reproduce locally) the error is not "File not found" but is failing on non-utf-8 characters:

```
Invalid argument(s): Illegal character identified in the token at col: 30
#0      StdUriTemplate._validateLiteral (file:///Users/aperuffo/workspace/std-uritemplate/dart/std_uri_template.dart:28:9)
#1      StdUriTemplate._expandImpl (file:///Users/aperuffo/workspace/std-uritemplate/dart/std_uri_template.dart:152:17)
#2      StdUriTemplate.expand (file:///Users/aperuffo/workspace/std-uritemplate/dart/std_uri_template.dart:7:12)
#3      main (file:///Users/aperuffo/workspace/std-uritemplate/dart/test.dart:29:35)
#4      _delayEntrypointInvocation.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:295:33)
#5      _RawReceivePort._handleMessage (dart:isolate-patch/isolate_patch.dart:184:12)
```

- do we need a "Verified publisher" I cannot register a GH Pages as verified: `std-uritemplate.github.io` as I don't control the DNS. Do you know if there is a way?
- can you update the readme and the documentation?
- can you draft the release GH Action?

**Comment by ricardoboss** [comment]
> fix the failure in CI (I can reproduce locally) the error is not "File not found" but is failing on non-utf-8 characters

Looking into it. I had similar issues running the tests locally, but shrugged it off hoping it was just a git issue on my machine.

> do we need a "Verified publisher" I cannot register a GH Pages as verified: std-uritemplate.github.io as I don't control the DNS. Do you know if there is a way?

AFAIK, that's the only way to create a verified publisher. It is possible to publish packages without being verified, but I wouldn't recommend that. Maybe the GitHub support can help? The pub.dev team could maybe also verify you?

> - can you update the readme and the documentation?
> - can you draft the release GH Action?

On it.

**Comment by ricardoboss** [comment]
Automated releases to pub.dev are possible after the first version has been manually published. I prepared the workflow, but I **strongly advise you** to read through this document: https://dart.dev/tools/pub/automated-publishing#configuring-automated-publishing-from-github-actions-on-pubdev

It not only describes the idea behind automated publishing, it also guides you how to set it up and harden its security.

**Comment by ricardoboss** [comment]
A curiosity about the the 'Language native date-time format' test: Dart implemented the ISO 8601 including the milliseconds, so instead of `2023-02-13T18:49:00Z` it results in `2023-02-13T18:49:00.000Z` (which is also ISO 8601). I added the formats including the milliseconds to the `edge-cases.json` as alternative results.

**Comment by ricardoboss** [comment]
@andreaTP all tests pass locally. I not sure whether I should redo the `StringBuffer` allocations. LMK if I should adjust anything, also happy to answer any questions or provide support regarding publishing!

File: dart/lib/std_uri_template.dart [hunk 2]
```diff
@@ -0,0 +1,517 @@
...
+        (0xF900 <= codePoint && codePoint <= 0xFDCF) ||
+        (0xFDF0 <= codePoint && codePoint <= 0xFFEF);
+  }
+
+  static void _addExpandedValue(dynamic prefix, dynamic value,
+      StringBuffer result, int maxChar, bool replaceReserved) {
+    final stringValue = _convertNativeTypes(value);
+    final runes = stringValue.runes;
+    final max = (maxChar != -1) ? min(maxChar, runes.length) : runes.length;
+    result.write('');
```
Comment (andreaTP) [inline dart/lib/std_uri_template.dart:295]: is this needed?

File: uritemplate-test-additional/edge-cases.json [hunk 3]
```diff
@@ -17,10 +17,10 @@
       "nativedatetwo": 1676314140000
      },
      "testcases" : [
-        ["{+nativedate}", "2023-02-13T18:49:00Z"],
-        ["{nativedate}", "2023-02-13T18%3A49%3A00Z"],
-        ["{+nativedatetwo}", "2023-02-13T18:49:00Z"],
-        ["{nativedatetwo}", "2023-02-13T18%3A49%3A00Z"]
+        ["{+nativedate}", ["2023-02-13T18:49:00Z", "2023-02-13T18:49:00.000Z"]],
```
Comment (andreaTP) [inline uritemplate-test-additional/edge-cases.json:20]: I know that this is another possible and valid output, although, to keep this implementation results predictable across implementation I have to ask you if it's possible to generate the proposed format (i.e. removing `.000`) using a DateTime Formatter or something similar.

**Review by andreaTP — CHANGES_REQUESTED** [review]
Thanks for bearing with me in this process @ricardoboss !
This looks almost good to merge, things left for you:

- please emit the date in the proposed format
- change the package name to `std_uritemplate` ( 🙏 )

For me:

- need to register the Verified Publisher, tentatively opened [this PR](https://github.com/proj-at/subdomains/pull/23) to have a valid domain for free

File: uritemplate-test-additional/edge-cases.json (hunk 3 above)
Comment (ricardoboss) [reply uritemplate-test-additional/edge-cases.json:20]: Sure that's possible! I'll revert this change and adjust the format.

File: dart/lib/std_uri_template.dart (hunk 2 above)
Comment (ricardoboss) [reply dart/lib/std_uri_template.dart:295]: Oops, oversight on my part. Should not be needed.

File: uritemplate-test-additional/edge-cases.json (hunk 3 above)
Comment (ricardoboss) [reply uritemplate-test-additional/edge-cases.json:20]: One thing though: Using `DateFormat` requires a dependency on the `intl` package. I think we both want to avoid dependencies. Should I add it anyway or resort to string manipulation to remove the milliseconds after converting the date to string?

Comment (andreaTP) [reply uritemplate-test-additional/edge-cases.json]: There should be no external dependencies, you are right, from what I see the best would be to have a self-contained function that does something similar to the standard library:

```dart
  String m = _twoDigits(month);
  String d = _twoDigits(day);
  String h = _twoDigits(hour);
  String min = _twoDigits(minute);
  String sec = _twoDigits(second);
  return "$y-$m-${d}T$h:$min:$secZ";
```

Comment (ricardoboss) [reply uritemplate-test-additional/edge-cases.json]: Done

File: dart/CHANGELOG.md [hunk 4]
```diff
@@ -0,0 +1,3 @@
+# 0.0.1
```
Comment (andreaTP) [inline dart/CHANGELOG.md:1]: According to [this](https://dart.dev/tools/pub/publishing#important-files) the Changelog is not strictly needed ...
This is not going to be maintained ...

File: dart/pubspec.yaml [hunk 5]
```diff
@@ -0,0 +1,14 @@
+name: std_uritemplate
+version: 0.0.1
```
Comment (andreaTP) [inline dart/pubspec.yaml:2]: We should update this version dynamically like we do, e.g. for typescript.

Something like this:

```
cat <<< $(yq e -M ".version = \"${GITHUB_REF_NAME}\"" pubspec.yaml) > pubspec.yaml
```
in the release workflow should make it ...

File: .github/workflows/release.yml [hunk 6]
```diff
@@ -196,3 +196,25 @@ jobs:
...
+    permissions:
+      id-token: write # Required for authentication using OIDC
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+      - name: Setup Dart
+        uses: dart-lang/setup-dart@v1
+        with:
+          sdk: stable
+      - name: Install dependencies
```
Comment (andreaTP) [inline .github/workflows/release.yml:211]: Should we run this command in the `dart` folder?

```
working-directory: dart
```

**Review by andreaTP — COMMENTED** [review]
A couple of minor asks again, but it looks good to me overall, thanks for the contribution!

Let me wait to see if I can register the subdomain, I'll try a first release from my machine, I'll merge the PR and we finalize a proper automated release.

File: dart/lib/std_uritemplate.dart [hunk 7]
```diff
@@ -0,0 +1,559 @@
+import 'dart:math';
+
+enum _SubstitutionType { EMPTY, STRING, LIST, MAP }
+
+enum _Operator { PLUS, HASH, DOT, SLASH, SEMICOLON, QUESTION_MARK, AMP, NO_OP }
+
+extension on DateTime {
+  /// This is the same as [toIso8601String] but without milli- and microseconds.
+  String toIso8601StringWithoutMilliseconds() {
```
Comment (andreaTP) [inline dart/lib/std_uritemplate.dart:9]: Implementation wise looks good, thanks for the change!
Please help me with Dart semantics, is this going to be public?
E.g. can a user use this function after importing the library?

Do we have a way to make it visible only to the current file/compilation unit?

Comment (ricardoboss) [reply dart/lib/std_uritemplate.dart:9]: No. In Dart, `extension`s are only accessible from outside the file if they have a name (i.e. `extension DateTimeExtension on DateTime {...}`). I am testing the library in another project and the extension method is not accessible.

File: .github/workflows/release.yml (hunk 6 above)
Comment (ricardoboss) [reply .github/workflows/release.yml:211]: Yes, good catch!

File: dart/pubspec.yaml (hunk 5 above)
Comment (ricardoboss) [reply dart/pubspec.yaml:2]: I used `sed` instead. Hope that's fine.

File: dart/CHANGELOG.md (hunk 4 above)
Comment (ricardoboss) [reply dart/CHANGELOG.md]: Removed

File: dart/lib/std_uritemplate.dart (hunk 7 above)
Comment (andreaTP) [reply dart/lib/std_uritemplate.dart:9]: perfect, thanks for the clarification 👍

**Comment by ricardoboss** [comment]
@andreaTP I think I addressed all comments. Can you review again?

File: dart/test.dart [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+        DateTime.fromMillisecondsSinceEpoch(data["nativedatetwo"]);
+  }
+
+  final String template;
+  try {
+    template = File(templateFile).readAsStringSync().trim();
+  } catch (e, stack) {
+    stderr.write("Error loading template file: $e\n$stack\n");
+
+    print("false");
```
Comment (andreaTP) [inline dart/test.dart:44]: Sorry if I missed this earlier, but this is incorrect, in case there are errors loading the file it's a "tool" issue and we should just explode.

File: dart/pubspec.yaml (hunk 5 above)
Comment (andreaTP) [reply dart/pubspec.yaml:2]: Just not ideal, have you experienced issues with my proposal?

File: dart/test.dart [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+  final templateFile = args[0];
+  final dataFile = args[1];
+
+  final dynamic data;
+  try {
+    data = jsonDecode(File(dataFile).readAsStringSync());
+  } catch (e, stack) {
+    stderr.write("Error loading data file: $e\n$stack\n");
+
+    print("false");
```
Comment (andreaTP) [inline dart/test.dart:23]: missed this as well, in this case, we should throw the exception and exit without printing a valid output.

**Review by andreaTP — CHANGES_REQUESTED** [review]
A couple of very last comments, I think this is good and basically ready to merge(in case I can fix things on top).
I'm waiting to merge only to see if I can get the sub-domain verified for free.

An alternative approach would be to comment out the "release" section, merge this PR and continue in a follow-up PR.
As you prefer 🙂 thanks a lot for this great contribution!

File: dart/pubspec.yaml (hunk 5 above)
Comment (ricardoboss) [reply dart/pubspec.yaml:2]: Yeah, the command you provided replaced the whole file content with just `version: ...`.  Maybe it is because I am using bash on Mac. I thought `sed` might be the better tool to achieve the task. But happy to adjust as you wish!

File: dart/test.dart (hunk 8 above)
Comment (ricardoboss) [reply dart/test.dart]: Now throwing

File: dart/test.dart (hunk 9 above)
Comment (ricardoboss) [reply dart/test.dart]: Fixed

**Comment by ricardoboss** [comment]
> An alternative approach would be to comment out the "release" section, merge this PR and continue in a follow-up PR.

I think I'd like that solution.

One other thing: should I add myself to the `CODEOWNERS` file for `*.dart`?

File: dart/pubspec.yaml (hunk 5 above)
Comment (andreaTP) [reply dart/pubspec.yaml:2]: Just tested on Mac with `yq` version `4.24.2` and it works as expected ... but ... well, is a nitpick, feel free to ignore me.

**Comment by andreaTP** [comment]
> I think I'd like that solution.

Ok, let's proceed in this way 👍 

> One other thing: should I add myself to the `CODEOWNERS` file for `dart/`?

Thanks for the proposal! Yes, please, be my guest 🙏 , much appreciated!

File: Readme.md [hunk 10]
```diff
@@ -26,6 +27,7 @@ This is intended to become a complete and maintained cross-language implementati
 | Ruby | ✅ | ❌ | ✅ |
 | PHP | ✅ | ✅ | ✅ |
 | Swift | ✅ | ❌ | ✅ |
+| Dart | ✅ | ❌ | ✅ |
```
Comment (andreaTP) [inline Readme.md:30]: I promise this is the last one, can you get an X on the "Releaded" column

**Review by andreaTP — APPROVED** [review]
Last nitpick and I invited you as a collaborator on this repo.
LGTM, great work!

File: Readme.md (hunk 10 above)
Comment (ricardoboss) [reply Readme.md:30]: Sorry, but I don't see this column 🙃  Did you mean "Reviewed"? That one already has an X, though. (I am not 100% sure what that means)

Comment (ricardoboss) [reply Readme.md:30]: Ohhhh published, yeah that makes more sense

Comment (andreaTP) [reply Readme.md]: yes, published 😅 

> I am not 100% sure what that means

We normally use it when, at least one expert in the specific language other than the author has reviewed the code.

**Comment by andreaTP** [comment]
A pleasure working with you @ricardoboss 🙏
